datastore-design/go-grpc-demo/storagebackend/timescaledb/findtimeseries.go

```diff
@@ -0,0 +1,127 @@
+package timescaledb
+
+import (
+	"datastore/datastore"
+	"fmt"
+
+	postgis "github.com/cridenour/go-postgis"
+	_ "github.com/lib/pq"
+)
+
+// matchAnyValCond creates the sub-condition of a WHERE clause that checks if string column 'name'
+// matches any of a set of values.
+// Returns the sub-condition to be appended to a WHERE clause.
+func matchAnyValCond(name string, values []string) string {
+	if len(values) == 0 {
+		return ""
+	}
+	inArg := ""
+	for i, value := range values {
+		if i > 0 {
+			inArg += ","
+		}
+		inArg += fmt.Sprintf("'%s'", value)
+	}
+	return fmt.Sprintf(" AND %s IN (%s)", name, inArg)
+}
+
+func equal(p1, p2 *datastore.Point) bool {
+	return (p1.Lat == p2.Lat) && (p1.Lon == p2.Lon)
+}
+
+// insidePolygonCond creates a sub-condition of a WHERE clause that checks if geo point column
+// 'name' is contained in a geo polygon.
+// Returns the sub-condition to be appended to a WHERE clause.
+func insidePolygonCond(name string, polygon0 *datastore.Polygon) (string, error) {
+
+	if polygon0 == nil {
+		// absent polygon => disable filtering (note: "" is equivalent to " AND TRUE")
+		return "", nil
+	}
+
+	points := polygon0.Points
+
+	// validate
+	if len(points) < 3 {
+		return "", fmt.Errorf("polygon must contain at least 3 points; found %d", len(points))
+	}
+	if equal(points[0], points[len(points) - 1]) {
+		return "", fmt.Errorf("polygon endpoints must be different")
+	}
+
+	// construct the polygon ring of the WKT representation
+	// (see https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry;
+	// note that only a single ring is supported for now)
+	polygonRing := ""
+	points = append(points, points[0]) // close polygon
+	for i, point := range points  {
+		if i > 0 {
+			polygonRing += ","
+		}
+		polygonRing += fmt.Sprintf("%f %f", point.Lon, point.Lat)
+	}
+
+	srid := "4326" // spatial reference system ID
+
+	return fmt.Sprintf(
+		" AND ST_WITHIN(ST_SetSRID(%s::geometry, %s), ST_GeomFromText('polygon((%s))', %s))",
+		name, srid, polygonRing, srid), nil
```

Comment:
Ok, will look into that.

Reply:
Pushed fix now (note that I didn't have to cast anything, only add the third arg).
I wonder though: Any spherical polygon (that is not self-intersecting) divides the sphere into two parts, so which of those parts is considered interior? Is it defined by "clockwiseness"?

Reply:
Final thought, it might be more clear to use [ST_Intersects](https://postgis.net/docs/en/ST_Intersects.html) (the geography version). So:
`" AND ST_Intersects(%s, ST_GeomFromText($%d, %s)::geography)"`

Reply:
Is the cartesian vs spherical issue really a tradeoff between performance and precision? In that case I would vote for precision, at least to begin with. I also feel that it is more natural to think in terms of a sphere and not worry about projections, singularities at poles etc.

Reply:
Regarding your thought about ST_INTERSECTS, this [post](https://gis.stackexchange.com/questions/227892/how-to-find-points-within-a-polygon-in-postgis) recommends ST_WITHIN over ST_INTERSECTS, but maybe it is not that relevant to our case? (must admit I'm not really a PostGIS expert as you probably might have guess by now :) )

Reply:
I just realized that I had misunderstood your earlier suggestion to cast the polygon (and didn't react on your follow-up comment either!). Pushed a fix that does `ST_GeomFromText(...)::geography` now.

Reply:
Well, for a bounding box, cartesian might be preferred. If your bounding box is:
left=5.0, bottom=50.0, right=15.0, top=60.0, would a user expect the following point to be in the box?
(10, 60.0001)
With cartesian it isn't, but with spherical it is (as the great circle between (5, 60) and (15,60) goes north of this point.

But for a random polygon, the great circle is what I would expect.
Another side effect of using spherical is that a bounding box of (-180, -90, 180, 90) becomes invalid, because that is actually a line between the north and south pole. I think postgis actually gives an error (I am on the wrong branch to test).

Anyway, I am happy with using ST_DWITHIN (the "D" is important), and everything spherical.

Reply:
Ok, good. I guess we will have to revisit this later anyway when we start to implement the EDR API. If we are required to support both bounding boxes and general polygons we can select the appropriate PostGIS function at runtime (and possibly even allowing the user to override the default behavior).

Reply:
Indeed, we can revisit.

Interestingly enough, we had exactly this discussion at our team in KNMI related to EDR.